packages/updating-element/src/updating-element.ts

```diff
@@ -930,37 +712,77 @@ export abstract class UpdatingElement extends HTMLElement {
     try {
       shouldUpdate = this.shouldUpdate(changedProperties);
       if (shouldUpdate) {
-        this._controllers?.forEach((c) => c.willUpdate?.(changedProperties));
         this.willUpdate(changedProperties);
-        this._controllers?.forEach((c) => c.update?.(changedProperties));
         this.update(changedProperties);
       } else {
-        this._markUpdated();
+        this._resolveUpdate();
       }
     } catch (e) {
       // Prevent `firstUpdated` and `updated` from running when there's an
       // update exception.
       shouldUpdate = false;
       // Ensure element can accept additional updates after an exception.
-      this._markUpdated();
+      this._resolveUpdate();
       throw e;
     }
     // The update is no longer considered pending and further updates are now allowed.
     if (shouldUpdate) {
-      this._didUpdate(changedProperties);
+      this.didUpdate(changedProperties);
     }
   }
 
-  willUpdate(_changedProperties: PropertyValues) {}
+  //@internal
+  _resolveUpdate() {
+    super._resolveUpdate();
+    this.isUpdatePending = false;
+  }
+
+  /**
+   * Controls whether or not `update` should be called when the element requests
+   * an update. By default, this method always returns `true`, but this can be
+   * customized to control when to update.
+   *
+   * @param _changedProperties Map of changed properties with old values
+   */
+  protected shouldUpdate(_changedProperties: PropertyValues): boolean {
+    return true;
+  }
+
+  /** @internal */
+  willUpdate(_changedProperties: PropertyValues) {
+    this._controllers?.forEach((c) => c.willUpdate?.());
+  }
+
+  /**
+   * Updates the element. This method reflects property values to attributes.
+   * It can be overridden to render and keep updated element DOM.
+   * Setting properties inside this method will *not* trigger
+   * another update.
+   *
+   * @param _changedProperties Map of changed properties with old values
+   * @internal
+   */
+  update(_changedProperties: PropertyValues) {
+    this._controllers?.forEach((c) => c.update?.());
+    if (this._reflectingProperties !== undefined) {
+      // Use forEach so this works even if for/of loops are compiled to for
+      // loops expecting arrays
+      this._reflectingProperties.forEach((v, k) =>
+        this._propertyToAttribute(k, this[k as keyof this], v)
+      );
+      this._reflectingProperties = undefined;
+    }
+    this._resolveUpdate();
+  }
 
   // Note, this is an override point for platform-support.
   // @internal
-  _didUpdate(changedProperties: PropertyValues) {
+  didUpdate(changedProperties: PropertyValues) {
+    this._controllers?.forEach((c) => c.didUpdate?.());
     if (!this.hasUpdated) {
       this.hasUpdated = true;
       this.firstUpdated(changedProperties);
     }
-    this._controllers?.forEach((c) => c.updated?.(changedProperties));
     this.updated(changedProperties);
```

Comment:
Will we eventually deprecate this in favor of `didUpdate`? Add a DEV_MODE warning to use `willUpdate` as `updated` may go away in a future version?

Reply:
Hmm, that would sort of leave `firstUpdated` as the only non-aligned name. I forget what we discussed about where to land with aligning the naming + where super is required.